File: companion/ios/Hearth/Core/HearthClient.swift
import Foundation

/// REST client for the Hearth pal-web FastAPI backend.
///
/// Design notes:
///   * Uses async/await; no third-party networking libraries.
///   * TLS pinning via `URLSessionDelegate` against the self-signed cert
///     captured during pairing.
///   * The `grantControl` method REQUIRES a `ConsentGesture`. That
///     parameter is not just for documentation — it is the only way to
///     obtain the `X-Consent-Origin: user-tap` header through
///     `ConsentTokenSource`. See `ConsentGesture.swift` for the type-level
///     enforcement.
actor HearthClient {
    nonisolated let baseURL: URL
    private let session: URLSession
    private let keychain: KeychainStore
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(baseURL: URL, keychain: KeychainStore = KeychainStore()) {
        self.baseURL = baseURL
        self.keychain = keychain

        let dec = JSONDecoder()
        dec.dateDecodingStrategy = .iso8601
        self.decoder = dec

        let enc = JSONEncoder()
        enc.dateEncodingStrategy = .iso8601
        self.encoder = enc

        let cfg = URLSessionConfiguration.ephemeral
        cfg.httpCookieAcceptPolicy = .always
        cfg.httpShouldSetCookies = true
        cfg.tlsMinimumSupportedProtocolVersion = .TLSv12
        cfg.waitsForConnectivity = false
        // Nothing leaves the house — do not use the URL cache.
        cfg.urlCache = nil
        cfg.requestCachePolicy = .reloadIgnoringLocalCacheData

        let ks = keychain
        let delegate = TLSPinningDelegate(pinProvider: { ks.data(for: .pinnedCert) })
        self.session = URLSession(configuration: cfg, delegate: delegate, delegateQueue: nil)
    }

    /// The URL session — exposed so companion views (WebSocket) inherit the pinning delegate.
    nonisolated var urlSession: URLSession { session }

    // MARK: - Auth

    struct LoginResponse: Decodable, Sendable {
        let csrfToken: String
        enum CodingKeys: String, CodingKey { case csrfToken = "csrf_token" }
    }

    func login(password: String) async throws -> LoginResponse {
        struct Body: Encodable { let password: String }
        let resp: LoginResponse = try await request(
            path: "/api/auth/login",
            method: "POST",
            body: Body(password: password),
            authenticated: false
        )
        try keychain.setString(resp.csrfToken, for: .csrf)
        return resp
    }

    // MARK: - Remote devices

    func devices() async throws -> [RemoteDevice] {
        try await request(path: "/api/remote/devices", method: "GET")
    }

    func pairDevice(name: String, kind: RemoteDevice.Kind) async throws -> RemoteDevice {
        struct Body: Encodable { let name: String; let kind: RemoteDevice.Kind }
        return try await request(
            path: "/api/remote/devices",
            method: "POST",
            body: Body(name: name, kind: kind)
        )
    }

    func unpair(deviceId: String) async throws {
        _ = try await requestRaw(
            path: "/api/remote/devices/\(deviceId)",
            method: "DELETE"
        )
    }

    // MARK: - Control grant (the load-bearing method)

    enum GrantError: Error, LocalizedError {
        case durationOutOfRange
        case cooldownActive
        case rollingCapExceeded
        case server(status: Int, body: String)

        var errorDescription: String? {
            switch self {
            case .durationOutOfRange: return "Duration must be between 1 and 60 minutes."
            case .cooldownActive:     return "Please wait 30 seconds between control grants."
            case .rollingCapExceeded: return "You have hit the 240-minute rolling 24h cap."
            case .server(let s, let b): return "Server \(s): \(b)"
            }
        }
    }

    /// Grant time-boxed control of `deviceId` to the Hearth.
    ///
    /// This method DEMANDS a `ConsentGesture`. The gesture parameter is
    /// the only path through which `ConsentTokenSource.header(for:)` can
    /// produce the `X-Consent-Origin: user-tap` header. See
    /// `ConsentGesture.swift` for the full argument.
    ///
    /// The method also verifies at runtime that the gesture's device id
    /// matches the `deviceId` we're posting to, catching mis-wired UIs.
    func grantControl(
        deviceId: String,
        gesture: ConsentGesture
    ) async throws -> ControlGrant {
        precondition(
            gesture.deviceId == deviceId,
            "ConsentGesture device id (\(gesture.deviceId)) does not match target (\(deviceId)). This is a programmer error in the calling view."
        )
        guard (1...60).contains(gesture.durationMinutes) else {
            throw GrantError.durationOutOfRange
        }

        struct Body: Encodable { let minutes: Int }
        let consent = ConsentTokenSource.header(for: gesture)
        let csrf = keychain.string(for: .csrf) ?? ""

        var extraHeaders: [String: String] = [
            consent.name: consent.value,
            "X-CSRF-Token": csrf,
        ]
        // Include a lightweight fingerprint of the gesture for the audit
        // log so a leaked or replayed body is distinguishable server-side.
        extraHeaders["X-Consent-Fingerprint"] = String(
            format: "%llx", UInt64(gesture.capturedAt.timeIntervalSince1970 * 1000)
        )

        do {
            return try await request(
                path: "/api/remote/devices/\(deviceId)/grant-control",
                method: "POST",
                body: Body(minutes: gesture.durationMinutes),
                extraHeaders: extraHeaders
            )
        } catch let ClientError.server(status, body) {
            switch status {
            case 409:  throw GrantError.cooldownActive
            case 429:  throw GrantError.rollingCapExceeded
            default:   throw GrantError.server(status: status, body: body)
            }
        }
    }

    func revokeControl(deviceId: String) async throws {
        _ = try await requestRaw(
            path: "/api/remote/devices/\(deviceId)/revoke-control",
            method: "POST"
        )
    }

    func openSession(deviceId: String) async throws -> RemoteSession {
        struct Body: Encodable { let device_id: String }
        return try await request(
            path: "/api/remote/sessions",
            method: "POST",
            body: Body(device_id: deviceId)
        )
    }

    func anomalies(sessionId: String) async throws -> [SessionAnomaly] {
        try await request(path: "/api/remote/sessions/\(sessionId)/anomalies", method: "GET")
    }

    // MARK: - Users / personality

    func users() async throws -> [HouseholdUser] {
        try await request(path: "/api/users", method: "GET")
    }

    func personality(userId: String) async throws -> PersonalityAxes {
        try await request(path: "/api/users/\(userId)/personality", method: "GET")
    }

    func updatePersonality(userId: String, axes: PersonalityAxes) async throws {
        _ = try await requestRaw(
            path: "/api/users/\(userId)/personality",
            method: "PUT",
            body: axes.clamped()
        )
    }

    // MARK: - Library / play

    func library() async throws -> [MediaItem] {
        try await request(path: "/api/library", method: "GET")
    }

    func play(itemId: String, target: PlayTarget) async throws {
        struct Body: Encodable { let item_id: String; let target_id: String }
        _ = try await requestRaw(
            path: "/api/play",
            method: "POST",
            body: Body(item_id: itemId, target_id: target.id)
        )
    }

    func extenders() async throws -> [Extender] {
        try await request(path: "/api/extenders", method: "GET")
    }

    // MARK: - Pairing

    struct PairResponse: Decodable, Sendable {
        let jwt: String
        let pinnedCert: String  // base64 DER
        enum CodingKeys: String, CodingKey {
            case jwt
            case pinnedCert = "pinned_cert"
        }
    }

    func pair(token: String, fingerprint: String) async throws -> PairResponse {
        struct Body: Encodable { let token: String; let fingerprint: String }
        let resp: PairResponse = try await request(
            path: "/api/pair",
            method: "POST",
            body: Body(token: token, fingerprint: fingerprint),
            authenticated: false
        )
        try keychain.setString(resp.jwt, for: .jwt)
        if let certData = Data(base64Encoded: resp.pinnedCert) {
            try keychain.set(certData, for: .pinnedCert)
        }
        return resp
    }

    // MARK: - Request plumbing

    enum ClientError: Error {
        case badURL
        case server(status: Int, body: String)
        case decoding(Error)
    }

    private func request<T: Decodable>(
        path: String,
        method: String,
        body: (some Encodable)? = Optional<Empty>.none,
        authenticated: Bool = true,
        extraHeaders: [String: String] = [:]
    ) async throws -> T {
        let (data, _) = try await requestRaw(
            path: path,
            method: method,
            body: body,
            authenticated: authenticated,
            extraHeaders: extraHeaders
        )
        do { return try decoder.decode(T.self, from: data) }
        catch { throw ClientError.decoding(error) }
    }

    @discardableResult
    private func requestRaw(
        path: String,
        method: String,
        body: (some Encodable)? = Optional<Empty>.none,
        authenticated: Bool = true,
        extraHeaders: [String: String] = [:]
    ) async throws -> (Data, HTTPURLResponse) {
        guard let url = URL(string: path, relativeTo: baseURL) else { throw ClientError.badURL }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")

        if authenticated, let jwt = keychain.string(for: .jwt) {
            req.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")
        }
        // Attach CSRF token to every mutating authenticated request.
        // grantControl passes its own X-CSRF-Token in extraHeaders; the loop
        // below runs AFTER this block so it wins.
        if authenticated, method != "GET" {
            if let csrf = keychain.string(for: .csrf), !csrf.isEmpty {
                req.setValue(csrf, forHTTPHeaderField: "X-CSRF-Token")
            }
        }
        for (k, v) in extraHeaders { req.setValue(v, forHTTPHeaderField: k) }

        if let body = body {
            req.httpBody = try encoder.encode(body)
        }

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw ClientError.server(status: -1, body: "not http")
        }
        guard (200..<300).contains(http.statusCode) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw ClientError.server(status: http.statusCode, body: text)
        }
        return (data, http)
    }

    private struct Empty: Encodable {}
}

/// Placeholder for the session response the WS handshake uses.
struct RemoteSession: Decodable, Identifiable, Sendable {
    let id: String
    let deviceId: String
    let wsPath: String
    enum CodingKeys: String, CodingKey {
        case id
        case deviceId = "device_id"
        case wsPath = "ws_path"
    }
}

// MARK: - TLS pinning delegate

/// Rejects the connection unless the server presents the exact cert we
/// captured during pairing. Falls back to the system trust store only if
/// we have not paired yet (during pairing itself, before the cert is in
/// the keychain). That transient window is the reason `pair()` is called
/// over the same session — the operator scans the QR to confirm the
/// fingerprint out of band.
final class TLSPinningDelegate: NSObject, URLSessionDelegate, URLSessionTaskDelegate, @unchecked Sendable {
    private let pinProvider: () -> Data?
    init(pinProvider: @escaping () -> Data?) { self.pinProvider = pinProvider }

    func urlSession(_ session: URLSession,
                    task: URLSessionTask,
                    didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        self.urlSession(session, didReceive: challenge, task: task, completionHandler: completionHandler)
    }

    func urlSession(_ session: URLSession,
                    didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        self.urlSession(session, didReceive: challenge, task: nil, completionHandler: completionHandler)
    }

    private func urlSession(_ session: URLSession,
                            didReceive challenge: URLAuthenticationChallenge,
                            task: URLSessionTask?,
                            completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        // Read the pin FRESH each challenge so a post-init pair() write is honored.
        let pinnedCert = pinProvider()
        // If we don't have a pinned cert yet, allow ONLY the local pairing path.
        // Any request to a path other than /api/pair on an unpinned socket is rejected.
        guard let pinnedCert else {
            let host = challenge.protectionSpace.host
            let path = task?.currentRequest?.url?.path ?? task?.originalRequest?.url?.path ?? ""
            let isLocalHost = host.hasSuffix(".local") || host == "127.0.0.1" || host == "localhost"
            let isPairPath = path == "/api/pair"
            if isLocalHost && isPairPath {
                completionHandler(.useCredential, URLCredential(trust: serverTrust))
            } else {
                completionHandler(.cancelAuthenticationChallenge, nil)
            }
            return
        }

        // Compare presented leaf cert against the pinned one.
        var certs: CFArray?
        if #available(iOS 15.0, *) {
            certs = SecTrustCopyCertificateChain(serverTrust)
        }
        let count = certs.map { CFArrayGetCount($0) } ?? 0
        guard count > 0,
              let leaf = (certs.flatMap { CFArrayGetValueAtIndex($0, 0) })
                  .map({ Unmanaged<SecCertificate>.fromOpaque($0).takeUnretainedValue() })
        else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        let leafData = SecCertificateCopyData(leaf) as Data
        if leafData == pinnedCert {
            completionHandler(.useCredential, URLCredential(trust: serverTrust))
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }
}

File: companion/android/app/src/main/kotlin/com/hearth/companion/core/HearthClient.kt
package com.hearth.companion.core

import com.hearth.companion.models.ControlGrant
import com.hearth.companion.models.Extender
import com.hearth.companion.models.MediaItem
import com.hearth.companion.models.PersonalityAxes
import com.hearth.companion.models.RemoteDevice
import com.hearth.companion.models.User
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.CertificatePinner
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import java.util.concurrent.TimeUnit

private val json = Json {
    ignoreUnknownKeys = true
    prettyPrint = false
    encodeDefaults = true
}

@Serializable
data class LoginRequest(val password: String)

@Serializable
data class LoginResponse(val csrf: String, val user: String)

@Serializable
data class PairRequest(val token: String, val device_name: String, val platform: String)

@Serializable
data class PairResponse(val jwt: String, val device_id: String, val spki_pin: String)

@Serializable
data class GrantRequest(val minutes: Int)

@Serializable
data class GrantResponse(val grant: ControlGrant)

@Serializable
data class SessionRequest(val device_id: String)

@Serializable
data class SessionResponse(val session_id: String, val ws_url: String)

@Serializable
data class PlayRequest(val item_id: String, val target_device_id: String)

@Serializable
data class DevicesResponse(val devices: List<RemoteDevice>)

@Serializable
data class UsersResponse(val users: List<User>)

@Serializable
data class LibraryResponse(val items: List<MediaItem>)

@Serializable
data class ExtendersResponse(val extenders: List<Extender>)

@Serializable
data class AnomalyLog(
    val session_id: String,
    val ts: String,
    val kind: String,
    val detail: String,
)

interface HearthApi {

    @POST("api/auth/login")
    suspend fun login(@Body body: LoginRequest): LoginResponse

    @POST("api/pair")
    suspend fun pair(@Body body: PairRequest): PairResponse

    @GET("api/remote/devices")
    suspend fun devices(): DevicesResponse

    @POST("api/remote/devices")
    suspend fun addDevice(@Body body: RemoteDevice, @Header("X-CSRF-Token") csrf: String): RemoteDevice

    @DELETE("api/remote/devices/{id}")
    suspend fun unpair(@Path("id") id: String, @Header("X-CSRF-Token") csrf: String)

    /**
     * The **only** Retrofit endpoint that accepts the X-Consent-Origin
     * header. Wired to route through [ConsentInterceptor] which asserts
     * the value must be exactly the string returned by [ConsentTokenSource].
     */
    @POST("api/remote/devices/{id}/grant-control")
    suspend fun grantControl(
        @Path("id") id: String,
        @Header("X-Consent-Origin") consentOrigin: String,
        @Header("X-Consent-Nonce") nonce: String,
        @Header("X-CSRF-Token") csrf: String,
        @Body body: GrantRequest,
    ): GrantResponse

    @POST("api/remote/devices/{id}/revoke-control")
    suspend fun revokeControl(@Path("id") id: String, @Header("X-CSRF-Token") csrf: String)

    @GET("api/remote/sessions/{id}/anomalies")
    suspend fun anomalies(@Path("id") id: String): List<AnomalyLog>

    @POST("api/remote/sessions")
    suspend fun openSession(@Body body: SessionRequest, @Header("X-CSRF-Token") csrf: String): SessionResponse

    @GET("api/users")
    suspend fun users(): UsersResponse

    @GET("api/users/{id}/personality")
    suspend fun personality(@Path("id") id: String): PersonalityAxes

    @PUT("api/users/{id}/personality")
    suspend fun updatePersonality(
        @Path("id") id: String,
        @Body body: PersonalityAxes,
        @Header("X-CSRF-Token") csrf: String,
    ): PersonalityAxes

    @GET("api/library")
    suspend fun library(): LibraryResponse

    @POST("api/play")
    suspend fun play(@Body body: PlayRequest, @Header("X-CSRF-Token") csrf: String)

    @GET("api/extenders")
    suspend fun extenders(): ExtendersResponse
}

/**
 * Belt-and-braces: even though the only Retrofit method that TAKES an
 * `X-Consent-Origin` header is [HearthApi.grantControl], we install an
 * OkHttp interceptor that strips the header from every OTHER request and
 * asserts on [HearthApi.grantControl] that the value came from
 * [ConsentTokenSource]. If a future refactor accidentally hardcodes
 * `"user-tap"` at any other call site, the request 400s at the client
 * before it hits the wire.
 */
private class ConsentInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val req = chain.request()
        val header = req.header("X-Consent-Origin")
        if (header != null) {
            val onGrantPath = req.url.encodedPath.endsWith("/grant-control")
            require(onGrantPath) {
                "X-Consent-Origin header present on non-grant path ${req.url.encodedPath}"
            }
            // In-app tests use `resetForTest` to clear the seen set; in
            // production the ConsentInterceptor cannot know the nonce is
            // fresh because ConsentTokenSource has already consumed it and
            // handed us the header value. So we just assert the shape.
            require(header == "user-tap") {
                "Only ConsentTokenSource may set X-Consent-Origin. got=$header"
            }
        }
        return chain.proceed(req)
    }
}

/**
 * Long-lived singleton. Reconfigured after pairing to install the SPKI pin.
 */
class HearthClient(
    baseUrl: String,
    spkiPin: String? = null,
    jwt: String? = null,
) {

    private val jwtInterceptor = Interceptor { chain ->
        val req = if (jwt != null) {
            chain.request().newBuilder().addHeader("Authorization", "Bearer $jwt").build()
        } else chain.request()
        chain.proceed(req)
    }

    private val ok: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(6, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .callTimeout(30, TimeUnit.SECONDS)
        .retryOnConnectionFailure(true)
        .addInterceptor(ConsentInterceptor())
        .addInterceptor(jwtInterceptor)
        .addInterceptor(HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BASIC })
        .apply {
            if (!spkiPin.isNullOrBlank()) {
                val host = baseUrl
                    .removePrefix("https://").removePrefix("http://")
                    .substringBefore('/').substringBefore(':')
                certificatePinner(
                    CertificatePinner.Builder()
                        .add(host, "sha256/$spkiPin")
                        .build()
                )
            }
        }
        .build()

    val api: HearthApi = Retrofit.Builder()
        .baseUrl(if (baseUrl.endsWith("/")) baseUrl else "$baseUrl/")
        .client(ok)
        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
        .build()
        .create(HearthApi::class.java)

    fun okHttp(): OkHttpClient = ok
}
